### DISCUSSION TITLE
Possibility for sub collections by index

### DISCUSSION THREAD

**Discussion by Gomez12** [ravendb/ravendb#18671] [Ideas]
Because Collections can still hold a lot of documents, and indexes are basically running sequentially through a complete collection I was wondering if it could be made possible to have some sort of SubCollections (defined by separate indexes)
where the SubCollection only contain links/pointers to the original documents.

So If I have like 100m products in a Collection Products, that I can define a subCollection productsInNewYork (which is just a simple .Where(location=="NewYork") so that if I define an index or any other action on it that it only has to go through 1M of Products instead of 100M.

Conceptually I think I could achieve this with a Map-Reduce index with an output Collection which simply does not reduce anything. But this will create a lot of IO and feels like computational to be very heavy as it would 100% duplicate the documents while basically I just want a filtered Collection with all the original documents.

**Comment by ayende** [comment]
What are you actually trying to achieve? Let's start from this. 
Do you have a lot of traffic on your indexes?
Note that indexes only apply to modified documents after their creation, so it doesn't matter how many items are there in the colelction.

**Reply by Gomez12** [reply]
I have a lot of modified documents every day, because we get product data / and/or reviews in bulk from an external source.

And I have like 10 indexes for New York. They are starting with a where condition.

Currently when I bulk import 1M products, then I have 10 indexes starting to go through 1M documents, which results in basically an IO storm (FS Cache helps in the beginning) of opening 10 x 1 Million documents, while only 100K may be for the NewYork indexes.
I was thinking that a subCollection could read through the 1M documents, and if only 100K matches the condition then I have 1M (subCollection) + 10x 100K (indexes) IO load.

Basically the biggest problem is the IO storm (/slow IO) which means the indexes start blocking themselves and the whole website starts degrading.
I can speed up the import (and minimise the degradation) by pausing indexes and later re-enabling them (one by one or in pairs) so they don't cause an IO storm, but I was thinking seeing that I have a lot of indexes which have the same precondition that maybe RavenDb could minimise the IO storm by only once selecting the precondition instead of for every index on every modified document.

**Reply by ayende** [reply]
How much memory do you have here? There should be a lot of memory reuse in such a scenario? 
Two indexes referencing the same docs will not load them twice.

**Reply by Gomez12** [reply]
Basically not enough :)

That is what I am saying with FS Cache, when the import starts everything goes fast as it reuses FS Cache (or ravendb cache) but as the import goes on and the indexes start to be at different parts of the import then it starts to slow down.
As index A can be at 50% of the import bulk, while index B is only at 5% (because it just does more work) and my memory can only hold 40% of the bulk in memory.

As some indexes are map-reduce with output that also cuts into the memory.
As we are also having regular use of the website that also cuts into the memory.

And the performance graph is also : Fast (while the import has enough memory as all indexes read the documents from memory)
Slow (as some indexes are slow and need to retrieve documents which are not in memory anymore)
Fast (as the fastest indexes complete and all the slow documents can retrieve basically the same documents from memory again)

Perhaps sharding will also solve the issue, because with an indexing process on each node it is not currently economically viable to up the memory on all nodes to receive a random bulk-import.

P.s. just as a side-note, although I would see value in the idea in general to minimise IO Storms, I have to say that I only started thinking about it when a bulk-import partner had an outage of a week so we had to process a huge backlog bulk-import which basically took down our site for several hours until we started manually juggling with indexes pausing / starting.
It happens during regular imports as well, but not on the scale that it is a huge problem.

It was just a thought on disaster recovery and just having the system chugging on instead of the website basically halting on insufficient memory which starts an IO storm.

**Reply by ayende** [reply]
Your issue is about capacity, because in the scenario you describe, you'll write 1M docs, then have an index process them and output another 100K? 
Then you'll read those 100K 10 times...

Versus reading from the disk (which is typically cheap). 

I assume that you are running on a cloud instance, which severely limits your IO.
 

You can set the import to insert only at a certain rate/second. May be easier to do things in that manner, to be honest. Slow things down. 
Alternatively, you can set indexes with `Indexing.Throttling.TimeIntervalInMs`, but I would recommend not going there.

Dripping the data in is probably the best case scenario.

**Reply by Gomez12** [reply]
> Your issue is about capacity, because in the scenario you describe, you'll write 1M docs, then have an index process them and > output another 100K?
> Then you'll read those 100K 10 times...

Basically that would be the scenario with a map-reduce output filter yes, why is was thinking about not outputting 100k docs, but just 100k ids.

>  You can set the import to insert only at a certain rate/second.
Do you mean this is RavenDb tech options somewhere? Or just regular rate-limiting.

> Alternatively, you can set indexes with Indexing.Throttling.TimeIntervalInMs, but I would recommend not going there.
Tried that, but either it didn't resolve much (on low values) or it just made larger io storms on larger values.

Probably no chance that you are going to put such a drip-feed tech on your roadmap?

Because we were thinking about a system like this, creating an import-db with no indexes and then drip-feed them over, but we ended up with pretty difficult to achieve metrics to determine the drip size because of caching.

We simply tried : 
Insert initial drip
Run query to see if response time is with a certain range.
Then according to query-time either halve, leave the same or double drip-size.

But we were hitting walls on caching / hot paths created for the query. And then mainly in Lucene / File System Cache (/systems out of control of the ravendb client).
But we will explore the drip-feeding further then by gathering more info from the server.

Or perhaps we will put upgrading to v6 and sharding further up the roadmap, as that would theoretically also solve it by just having smaller indexes per shard.

**Reply by ayende** [reply]
How are you importing the data? 

And what kind of hardware are you running on? 

What is the total size of the system?
